### ISSUE TITLE
Ability to select how a bibliography is sorted

### ISSUE THREAD

**Issue by MichalNemecek** [typst/typst#3913]
### Description

When creating a bibliography and citing the sources from it in the article, the generated bibliography is sorted in the order in which the first citations of each source appears in the document. However, this isn't always desired. I propose adding a `sort` parameter to  the bibliography function to allow sorting in a different way, possibly even the ability to use a function for sorting.

### Use Case

Like I mentioned in the description, you might not want the sources sorted based on the first appearances in the document. For example, my high school graduation thesis guidelines dictate that every bibliography entry should be cited, either in the text or as a footnote (thus shuffling around the bibliography order), but the bibliography at the end should be sorted alphabetically.

For the time being, I came up with a temporary solution: there's a huge blank area on the 2nd (statutory declaration) and 3rd (acknowledgement) page, so I just cited the bibliography in a `hide` block in small print, so as to not disturb the layout of the page in case the number of sources grows. That way, I can manually sort them in any order I need.

**Comment by EpicEricEE** [comment]
As a workaround if you have access to the CSL file, you can add
```xml
<sort>
  <key variable="citation-label" />
</sort>
```
inside the `bibliography` section to sort it by the label (e.g. `Cur98`, `Ein05`).

**Comment by laurmaedje** [comment]
There could be an easier way I guess, but I wouldn't really call this a workaround, but rather the intended solution. Generally, the sorting behaviour is mandated by the bibliography style.

**Comment by Andrew15-5** [comment]
I have very strange guidelines (or maybe the lack of them). Normally I would sort bibliography items alphabetically (by author, title, and year). But in a provided example I have, the term "order" does not exit. It's just some random order in which items were added. So, "logically" speaking, I have to put my (additional) items to the end of the list. In other words, the bibliography has to mirror the order in which items are defined in a bibliography file. And as soon as I cite something, it will jump to the top of the list. The `citation-label` was _almost_ what I wanted. Thankfully I found `citation-key`, which did what I wanted:

```xml
<bibliography>
    <sort>
      <key variable="citation-key" />
    </sort>
...
```

Now, I don't understand what is the default sorting algorithm when `<sort>` doesn't exist. I would assume that it's `citation-key`, which it is, until you start citing something in the document. Can you really make a "first place items that were cited in the order of citing, then place uncited items in the order they are defined"? But in any case, perhaps overriding the default sorting can be nice, because otherwise you have to understand CSL and change the style file, and if you don't have it locally, you have to also download it and store it somewhere.

Although this is probably going to be a part of general bibliography customization: #942.

**Comment by tobiasdiez** [comment]
Especially in the context of CV / publication lists, it may be helpful to sort (+ group) references differently than what the citation style might prescribe.

**Comment by johentsch** [comment]
> Generally, the sorting behaviour is mandated by the bibliography style.

That's right and still, biblatex gives you full control over the sort order (as in `\usepackage[sorting=ydnt]{biblatex}`) precisely for the reason there are many settings where another sort order is required, the most common of which, as @tobiasdiez indicated, (reverse) chronological order.

**Comment by Retengart** [comment]
https://www.zotero.org/styles/?q=gost

<img width="800" height="300" alt="Image" src="https://github.com/user-attachments/assets/bffc8195-b787-4829-b7f4-346692c70059" />
